src/v/cloud_storage/segment_chunk_api.cc

```diff
@@ -0,0 +1,345 @@
+                to_release.push_back(it);
+            }
+        }
+    }
+
+    bool need_trim = hydrated_chunks > _max_hydrated_chunks;
+    vlog(
+      _ctxlog.trace,
+      "{} hydrated chunks, need trim: {}",
+      hydrated_chunks,
+      need_trim);
+    if (!need_trim) {
+        co_return;
+    }
+
+    std::sort(
+      to_release.begin(),
+      to_release.end(),
+      [](const auto& it_a, const auto& it_b) {
+          return it_a->second <=> it_b->second == std::strong_ordering::less;
+      });
+
+    std::vector<ss::lw_shared_ptr<ss::file>> files_to_close;
+    files_to_close.reserve(to_release.size());
+
+    // The files to close are first moved out of chunks and into a vector. This
+    // is necessary to make sure that the chunks are no longer able to use the
+    // file handle about to be closed before a scheduling point. If the file
+    // handles are closed in this loop, because the close operation is
+    // a scheduling point, it is possible that a chunk may get its file handle
+    // acquired by a reader while it is waiting to be closed.
+    for (auto& it : to_release) {
+        if (hydrated_chunks <= _max_hydrated_chunks) {
+            break;
+        }
+
+        vlog(
+          _ctxlog.trace,
+          "marking chunk starting at offset {} for release, pending for "
+          "release: {} chunks",
+          it->first,
+          hydrated_chunks - _max_hydrated_chunks);
+        files_to_close.push_back(std::move(it->second.handle.value()));
+        it->second.handle = std::nullopt;
+        it->second.current_state = chunk_state::not_available;
+        hydrated_chunks -= 1;
+    }
+
+    std::vector<ss::future<>> fs;
+    fs.reserve(files_to_close.size());
+
+    for (auto& f : files_to_close) {
+        fs.push_back(f->close());
+    }
+
+    auto close_results = co_await ss::when_all(fs.begin(), fs.end());
+    for (const auto& result : close_results) {
+        if (result.failed()) {
+            vlog(
+              _ctxlog.warn,
+              "failed to close a chunk file handle during eviction");
+        }
+    }
+
+    _eviction_timer.rearm(_eviction_jitter());
+}
+
+void segment_chunks::register_readers(
+  chunk_start_offset_t first, chunk_start_offset_t last) {
+    vassert(_started, "chunk API is not started");
+
+    if (last <= first) {
+        // If working with a single chunk, there are no future readers.
+        return;
+    }
+
+    auto start = _chunks.find(first);
+    vassert(
+      start != _chunks.end(), "No chunk found starting at first: {}", first);
+
+    auto end = _chunks.find(last);
+    vassert(end != _chunks.end(), "No chunk found starting at last: {}", last);
+
+    auto required_after = 1;
+    end = std::next(end);
+    for (auto it = start; it != end; ++required_after, ++it) {
+        auto& chunk = it->second;
+        chunk.required_by_readers_in_future += 1;
+        chunk.required_after_n_chunks += required_after;
```

Comment:
This would usually be taken care of by the higher count of `required_by_readers_in_future` for overlapping chunk which is the primary property during sort, but it does seem like it could be a problem if `required_by_readers_in_future` is equal for multiple chunks. I will look into updating it.

Reply:
Since the default chunk trimming mode is `eager`, I'd say it's fine to do this in a follow up.